pysixdesk/lib/preprocess.py

```diff
@@ -1,366 +1,385 @@
+        finally:
+            # if failed and mysql or if success --> store results
+            # result store location
+            if self.db_type == 'mysql':
+                dest_path = Path('./result')
+            else:
+                dest_path = Path(self.madx_cfg["dest_path"]) / str(self.task_id)
+            if not dest_path.is_dir():
+                dest_path.mkdir(parents=True, exist_ok=True)
+            # download results
+            self.dl_output(dest_path)
+            # push results to mysql db
+            if self.db_type == 'mysql':
+                if self.coll_flag:
+                    self.madx_out['fort3.limi'] = 'fort3.limi'
+                self.push_to_db(dest_path)
+
+    def madx_copy_mask(self):
+        '''
+        Copies madx mask file to cwd.
+        '''
+        mask_name = self.madx_cfg["mask_file"]
+        source_path = Path(self.madx_cfg['source_path'])
+        shutil.copy2(source_path / mask_name, mask_name)
+        # make destination folder
+        Path(self.madx_cfg['dest_path']).mkdir(parents=True, exist_ok=True)
+
+    def madx_prep(self, output_file='madx_in'):
+        '''Replaces the placeholders in the mask_file and prints diff.
+
+        Args:
+            output_file (str, optional): Name of the prepared mask_file.
+
+        Raises:
+            Exception: if failure to generate mask file.
+        '''
+        patterns = ['%' + a for a in self.mask_cfg.keys()]
+        values = list(self.mask_cfg.values())
         try:
-            utils.replace(patterns, values, source, dest)
+            utils.replace(patterns,
+                          values,
+                          self.madx_cfg["mask_file"],
+                          output_file)
         except Exception:
-            content = "Failed to generate input file for oneturn sixtrack!"
+            content = "Failed to generate actual madx input file!"
+            raise Exception(content)
+
+        # show diff
+        # mask_name = self.madx_cfg["mask_file"]
+        # utils.diff(mask_name, output_file, logger=self._logger)
+
+    def madx_run(self, mask):
+        """Runs madx.
+
+        Args:
+            mask (str): mask file on which to run madx.
+
+        Raises:
+            Exception: If 'finished normally' is not in Madx output.
+        """
+        exe = self.madx_cfg['madx_exe']
+        command = exe + " " + mask
+        self._logger.info("Calling madx %s" % exe)
+        self._logger.info("MADX job is running...")
+        output = os.popen(command)
+        output = output.readlines()
+        with open('madx_stdout', 'w') as mad_out:
+            mad_out.writelines(output)
+        if 'finished normally' not in output[-2]:
+            content = "MADX has not completed properly!"
             raise Exception(content)
+        else:
+            self._logger.info("MADX has completed properly!")
+
+    def madx_job(self):
+        """
```

Comment:
It should not be about how many lines a function is. These 2 methods exists in order to separate functional from structural. With these 2 functions it helps to make explicit the execution logic of the job contained in the `run` method, and makes it very easy to change the job's logic. I don't see any disadvantages to having them.

And also, if we end up going the class mixin direction, as we discussed, having these methods makes it easier.

Reply:
If we split a job into several steps, and each step with a method. Of course, it will be more flexible. But do we really need this flexibility?  For the current case, it's unnecessary for the madx job, the workflow is simple and complete. Even we use mixin feature, there is no case to change the workflow of madx job, if there is suddenly, I think we should modify the workflow directly rather than using mixin. So it's not about disadvantage, the changes make no sense. 
As I said, we should avoid making unnecessary changes.